FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -0,0 +1,251 @@
+  }
+
+  /// Build custom model object from model info.
+  func buildModel() -> CustomModel? {
+    /// Build custom model only if model info is filled out, and model file is already on device.
+    guard let info = modelInfo, let path = info.path else { return nil }
+    let model = CustomModel(
+      name: info.name,
+      size: info.size,
+      path: path,
+      hash: info.hash
+    )
+    return model
+  }
+}
+
+/// Extension to handle fetching model info from server.
+extension ModelInfoRetriever {
+  /// HTTP request headers.
+  static let fisTokenHTTPHeader: String = "X-Goog-Firebase-Installations-Auth"
+  static let hashMatchHTTPHeader: String = "If-None-Match"
+  static let bundleIDHTTPHeader: String = "X-Ios-Bundle-Identifier"
+
+  /// HTTP response headers.
+  static let etagHTTPHeader: String = "ETag"
+
+  /// Error descriptions.
+  static let tokenErrorDescription: String = "Error retrieving FIS token."
+  static let modelFetchURLErrorDescription: String = "Error retrieving model fetch URL."
+  static let missingModelHashErrorDescription: String = "Model hash missing in server response."
+  static let invalidHTTPResponseErrorDescription: String =
+    "Could not get a valid HTTP response from server."
+
+  /// Construct model fetch base URL.
+  var modelInfoFetchURL: URL {
+    var components = URLComponents()
+    components.scheme = "https"
+    components.host = "firebaseml.googleapis.com"
+    components.path = "/Model/v1beta2/projects/\(projectID)/models/\(modelName)"
+    return components.url!
+  }
+
+  /// Construct model fetch URL request.
+  func getModelInfoFetchURLRequest(token: String) -> URLRequest {
+    var request = URLRequest(url: modelInfoFetchURL)
+    request.httpMethod = "GET"
+    // TODO: Check if bundle ID needs to be part of the request header.
+    let bundleID = Bundle.main.bundleIdentifier!
+    request.setValue(bundleID, forHTTPHeaderField: ModelInfoRetriever.bundleIDHTTPHeader)
+    request.setValue(token, forHTTPHeaderField: ModelInfoRetriever.fisTokenHTTPHeader)
+    if let info = modelInfo, info.hash.count > 0 {
+      request.setValue(info.hash, forHTTPHeaderField: ModelInfoRetriever.hashMatchHTTPHeader)
+    }
+    return request
+  }
+
+  /// Get model info from server.
+  func downloadModelInfo(completion: @escaping (DownloadError?) -> Void) {
+    /// Get FIS token.
+    installations.authToken { [weak self] tokenResult, error in
+      guard let result = tokenResult
+      else { completion(.internalError(description: ModelInfoRetriever.tokenErrorDescription))
+        return
+      }
+      /// Get model info fetch URL with appropriate HTTP headers.
+      guard let request = self?.getModelInfoFetchURLRequest(token: result.authToken)
+      else {
+        completion(.internalError(description: ModelInfoRetriever.modelFetchURLErrorDescription))
+        return
+      }
+      /// Download model info.
+      let dataTask = URLSession.shared.dataTask(with: request) { [weak self]
```

Comment:
nit: will it be more readable if we move sending the request to a separate method?

Reply:
True. I've added a TODO for myself to do this in a future PR, since this one is primarily about user defaults and FIS.